alerter/src/main.rs

```diff
@@ -0,0 +1,248 @@
+    }
+
+    async fn update_open_orders(&mut self) -> Result<()> {
+        let mut orders = self
+            .orderbook_api
+            .solvable_orders()
+            .await
+            .context("solvable_orders")?;
+        tracing::debug!("found {} open orders", orders.len());
+        std::mem::swap(&mut self.open_orders, &mut orders);
+        // Keep only orders that were open last update and are not open this update.
+        orders.retain(|order| !self.open_orders.contains(order));
+        for closed_order in orders {
+            let order = self
+                .orderbook_api
+                .order(&closed_order.order_meta_data.uid)
+                .await?;
+            if order.order_meta_data.status == OrderStatus::Fulfilled {
+                tracing::debug!(
+                    "updating last observed trade because order {} was fulfilled",
+                    order.order_meta_data.uid
+                );
+                self.last_observed_trade = Instant::now();
+                break;
+            }
+        }
+        tracing::debug!("found no fulfilled orders");
+        Ok(())
+    }
+
+    fn order_has_minimum_age(&self, order: &Order) -> bool {
+        let order_time =
+            Duration::from_secs(order.order_meta_data.creation_date.timestamp() as u64);
+        let now = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap();
+        now.saturating_sub(order_time) > self.config.min_order_age
+    }
+
+    fn alert(&self, order: &Order) {
+        tracing::error!(
+            "No orders have been settled in the last {} seconds even though order {} is solvable and has a price that allows it to be settled according to matcha.",
+            self.config.time_without_trade.as_secs(),
+            order.order_meta_data.uid,
+        );
+    }
+
+    pub async fn update(&mut self) -> Result<()> {
+        self.update_open_orders().await?;
+        if self.last_observed_trade.elapsed() <= self.config.time_without_trade {
+            return Ok(());
+        }
+        if matches!(
+            self.last_alert,
+            Some(instant) if instant.elapsed() < self.config.min_alert_interval
+        ) {
+            return Ok(());
+        }
+        for order in &self.open_orders {
+            if !self.order_has_minimum_age(order) {
+                continue;
+            }
+
+            if self
+                .matcha_api
+                .can_be_settled(&order.order_creation)
+                .await
+                .context("can_be_settled")?
+            {
+                self.last_alert = Some(Instant::now());
+                self.alert(order);
+                break;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[tokio::main]
+async fn main() {
+    let time_format_string = "%Y-%m-%dT%H:%M:%S%.3fZ";
+    tracing_subscriber::fmt::fmt()
+        .with_timer(tracing_subscriber::fmt::time::ChronoUtc::with_format(
+            String::from(time_format_string),
```

Comment:
General Question: Is there a reason you chose to cast to string only inside the constructor here rather than initializing `time_format_string` immediately as with `String::from`?

Reply:
Functionally it doesn't matter. Just slightly easier to read this way. I copied this code from shared/src/tracing.rs